go/pkg/rexec/rexec.go

```diff
@@ -332,48 +336,65 @@ func (ec *Context) ExecuteRemotely() {
 	ec.Metadata.RealBytesUploaded = bytesMoved
 	log.V(1).Infof("%s %s> Executing remotely...\n%s", cmdID, executionID, strings.Join(ec.cmd.Args, " "))
 	ec.Metadata.EventTimes[command.EventExecuteRemotely] = &command.TimeInterval{From: time.Now()}
-	eg, ctx := errgroup.WithContext(ec.ctx)
+	// Initiate each streaming request once at most.
 	var streamOut, streamErr sync.Once
-	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ctx, &repb.ExecuteRequest{
+	var streamWg sync.WaitGroup
+	// These variables are owned by the progress callback (which is async but not concurrent) until the execution returns.
+	var nOutStreamed, nErrStreamed int64
+	op, err := ec.client.GrpcClient.ExecuteAndWaitProgress(ec.ctx, &repb.ExecuteRequest{
 		InstanceName:    ec.client.GrpcClient.InstanceName,
 		SkipCacheLookup: !ec.opt.AcceptCached || ec.opt.DoNotCache,
 		ActionDigest:    ec.Metadata.ActionDigest.ToProto(),
 	}, func(md *repb.ExecuteOperationMetadata) {
 		if !ec.opt.StreamOutErr {
 			return
 		}
+		// The server may return either, both, or neither of the stream names, and not necessarily in the same or first call.
+		// The streaming request for each must be initiated once at most.
 		if name := md.GetStdoutStreamName(); name != "" {
 			streamOut.Do(func() {
-				eg.Go(func() error {
+				streamWg.Add(1)
+				go func() {
 					path := fmt.Sprintf("%s/logstreams/%s", ec.client.GrpcClient.InstanceName, name)
 					log.V(1).Infof("%s %s> Streaming to stdout from %q", cmdID, executionID, path)
-					_, err = ec.client.GrpcClient.ReadResourceTo(ctx, path, outerr.NewOutWriter(ec.oe))
-					return err
-				})
+					// Ignoring the error here since the net result is downloading the full stream after the fact.
+					n, err := ec.client.GrpcClient.ReadResourceTo(ec.ctx, path, outerr.NewOutWriter(ec.oe))
+					if err != nil {
+						log.Errorf("%s %s> error streaming stdout: %v", cmdID, executionID, err)
+					}
+					nOutStreamed += n
+					streamWg.Done()
```

Comment:
Nit: it's a bit more robust if you use `defer wg.Done()` as the first statement in the goroutine. That way some later change that exits the goroutine early won't result in Done() not being called (which would cause a deadlock).